sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -603,7 +603,14 @@ case class ExternalMapToCatalyst private(
 
   override def foldable: Boolean = false
 
-  override def dataType: MapType = MapType(keyConverter.dataType, valueConverter.dataType)
+  override def dataType: MapType = {
+    val isPrimitiveType = valueType match {
+      case BooleanType | ByteType | ShortType | IntegerType | LongType |
+            FloatType | DoubleType => true
+      case _ => false
+    }
+    MapType(keyConverter.dataType, valueConverter.dataType, !isPrimitiveType)
```

Comment:
Actually we already do the similar thing at JavaTypeInference https://github.com/apache/spark/blob/f00df40cfefef0f3fc73f16ada1006e4dcfa5a39/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala#L107.

Looks like we can infer value nullability at the places we create `ExternalMapToCatalyst` more precisely. It would not be a large change with the current type inference utils.